### PR TITLE
deploy full ESM with "type": "module" in "latest-esm" npm tag

### PR DESCRIPTION
**Pull request by PabloSzx**
I also just re-published graphql@16.0.1 as graphql-esm@16.0.1 https://npm.im/graphql-esm, tested it in an example project using libraries with full ESM support https://github.com/PabloSzx/test-graphql-esm, everything works perfectly 👌 

It can be tested right now doing this:
```json
{
  "dependencies": {
    "graphql": "npm:graphql-esm@^16.0.1"
  }
}
```

### CODE REVIEW

File: resources/build-npm-esm.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
+'use strict';
```
Comment (IvanGoncharov) [inline resources/build-npm-esm.js:1]: since most of the code is shared with `build-npm.js` it makes sense to just add a cli option to it.
Also please check all places where `build-npm` is used directly or through NPM cli and adjust them accordingly.

**Review by IvanGoncharov — COMMENTED** [review]
Please add `integrationTests/esm` for this package.
Try to test as many esm-specific things as possible.

**Comment by IvanGoncharov** [comment]
Also please ensure that this package works with upcoming TS 4.5 that should support esm natively and add integration tests for that.

File: integrationTests/ts-esm/package.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  "private": true,
+  "description": "graphql-js ESM should compile with all supported TS versions",
+  "scripts": {
+    "test": "node test.js"
+  },
+  "dependencies": {
+    "graphql": "file:../graphql-esm.tgz",
+    "typescript-4.2": "npm:typescript@4.2.x",
```
Comment (PabloSzx) [inline integrationTests/ts-esm/package.json:9]: For some reason Typescript@4.1.x didn't work, I'm not completely sure the reason :man_shrugging:

File: resources/build-npm.js [hunk 3]
```diff
@@ -8,30 +8,50 @@ const ts = require('typescript');
...
+    const destPath = path.join(distDirectory, filepath);
 
     fs.mkdirSync(path.dirname(destPath), { recursive: true });
     if (filepath.endsWith('.ts')) {
-      const cjs = babelBuild(srcPath, { envName: 'cjs' });
-      writeGeneratedFile(destPath.replace(/\.ts$/, '.js'), cjs);
-
-      const mjs = babelBuild(srcPath, { envName: 'mjs' });
-      writeGeneratedFile(destPath.replace(/\.ts$/, '.mjs'), mjs);
+      if (isFullESM) {
```
Comment (IvanGoncharov) [inline resources/build-npm.js:40]: Please rename it to `isESMOnly`

File: resources/build-npm-esm.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+const { buildNPM } = require('./build-npm.js');
+
+if (require.main === module) {
+  buildNPM({
+    isFullESM: true,
+  });
+}
```
Comment (IvanGoncharov) [inline resources/build-npm-esm.js:9]: please use CLI flag (e.g. `--esm-only`) to `build-npm`

File: integrationTests/webpack/webpack.config.json [hunk 5]
```diff
@@ -2,6 +2,8 @@
   "mode": "production",
   "entry": "./entry.js",
   "output": {
-    "libraryTarget": "commonjs2"
+    "library": {
+      "type": "commonjs2"
+    }
```
Comment (IvanGoncharov) [inline integrationTests/webpack/webpack.config.json:7]: Is it a fix?
If so please extract it in a separate PR and provide an explanation for it.

File: integrationTests/ts/package.json [hunk 6]
```diff
@@ -9,6 +9,8 @@
     "typescript-4.1": "npm:typescript@4.1.x",
     "typescript-4.2": "npm:typescript@4.2.x",
     "typescript-4.3": "npm:typescript@4.3.x",
-    "typescript-4.4": "npm:typescript@4.4.x"
+    "typescript-4.4": "npm:typescript@4.4.x",
+    "typescript-4.5": "npm:typescript@4.5.1-rc",
+    "typescript-next": "npm:typescript@next"
```
Comment (IvanGoncharov) [inline integrationTests/ts/package.json:14]: please don't do it next, if TS releases a new version it will break CI.

File: integrationTests/ts-esm/package.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (IvanGoncharov) [inline integrationTests/ts-esm/package.json:1]: AFAIK, ESM changes only imports, it doesn't change any other aspect.
Please don't duplicate the entire test folder for it.
Just create a separate ESM-specific test file and alias ESM packages as `graphql-esm`.

Comment (IvanGoncharov) [reply integrationTests/ts-esm/package.json:1]: Please do the same for webpack

File: integrationTests/node-esm/index.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable node/no-missing-import, import/no-unresolved */
```
Comment (IvanGoncharov) [inline integrationTests/node-esm/index.js:1]: Please don't just duplicate `node` test here.
Please try to test ESM-specific things here (e.g. importing `package.json`, dynamic import, etc.).
Also please create a test setup where you have an intermediate package with `peerDepenedicies` for `graphql` and test how it works with ESM package.

File: package.json [hunk 9]
```diff
@@ -56,6 +56,7 @@
     "@babel/plugin-transform-typescript": "7.15.8",
     "@babel/preset-env": "7.15.8",
     "@babel/register": "7.15.3",
+    "@types/babel__core": "^7.1.16",
```
Comment (IvanGoncharov) [inline package.json:59]: Please explain this change?
Please do that for other non-straight forward changes, beforehand.

File: integrationTests/webpack/webpack.config.json (hunk 5 above)
Comment (PabloSzx) [reply integrationTests/webpack/webpack.config.json:7]: image


I will remove it, but I won't make a separate PR just for

File: integrationTests/ts/package.json (hunk 6 above)
Comment (PabloSzx) [reply integrationTests/ts/package.json:14]: > please don't add next, if TS releases a new version it will break CI.

No problem, but I was just doing what you told me to, keep in mind that the new typescript ESM is no longer in 4.5.1, it's in @ next, see https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-rc/#beta-delta

File: package.json (hunk 9 above)
Comment (PabloSzx) [reply package.json:59]: image

vs

image

Comment (IvanGoncharov) [reply package.json]: Yes, it's a JS file, not TS.
Converting those files to TS is not part of this PR.

File: integrationTests/webpack/webpack.config.json (hunk 5 above)
Comment (IvanGoncharov) [reply integrationTests/webpack/webpack.config.json]: It's totally worth it, same as any other unrelated change.
Every change that makes sense on its own should be a separate PR.

File: package.json (hunk 9 above)
Comment (PabloSzx) [reply package.json]: it doesn't have to be TS for it to have autocompletion, and this change was not making this file typescript, but whatever you want 👍

Comment (saihaj) [reply package.json]: > Converting those files to TS is not part of this PR.

don’t think anything is changed. adding @types shouldn’t be an issue it just helps improve DX in vscode at least.

File: .gitignore [hunk 10]
```diff
@@ -14,3 +14,4 @@
 /denoDist
 /npm
 /deno
+/npmEsmDist
```
Comment (IvanGoncharov) [inline .gitignore:17]: Can you please move it to the same place as in `.eslintignore`?
All ignore files should have the same order.

File: resources/gen-version.js [hunk 11]
```diff
@@ -4,21 +4,31 @@ const fs = require('fs');
 
 const { version } = require('../package.json');
 
-const versionMatch = /^(\d+)\.(\d+)\.(\d+)-?(.*)?$/.exec(version);
-if (!versionMatch) {
-  throw new Error('Version does not match semver spec: ' + version);
+if (require.main === module) {
```
Comment (IvanGoncharov) [inline resources/gen-version.js:7]: @PabloSzx Please revert, we shouldn't run `npm version` twice and create two sets of tags and duplicate other steps.
It just that `npm run build:npm` should now produce two folder standard and ESM.
And `16.0.1` is `16.0.1` even if it's published as ESM.
If devs for some reason want to know if it's ESM or not they should import `package.json`

**Review by IvanGoncharov — CHANGES_REQUESTED** [review]

File: package.json (hunk 9 above)
Comment (IvanGoncharov) [reply package.json]: @saihaj Yes, it makes sense, but as separate PR.
It has nothing to do with ESM specifically.

Every commit/PR should be topic-based and revertable.
Also, it doesn't help if `git blame` shows that this package was added for the ESM package.

DX improvements can be done but they should be done in separate PRs with explanations not as random changes on unrelated PRs

**Review by yaacovCR — APPROVED** [review]
Looks good to me!

looks like esm tag makes sense out of band from canaries and all that

**Comment by yaacovCR** [comment]
> Looks good to me!
> 
> looks like esm tag makes sense out of band from canaries and all that

@IvanGoncharov  is this ready for re-review?

**Review by saihaj — APPROVED** [review]

**Comment by IvanGoncharov** [comment]
@github-actions publish-pr-on-npm

**Comment by IvanGoncharov** [comment]
I rebased this PR to the latest main as the solution for #3603

**Comment by yaacovCR** [comment]
@IvanGoncharov @PabloSzx  et al

question about the current state => the build scripts add a build qualifier of "+esm" and a tag suffix of "-esm", but do not rename the package to `graphql-esm` => the integration tests that use `graphql-esm` just point to a local tarball with the distribution, presumably where the name of the package is still `graphql`.

Over at the `graphql-esm` on npm -- it doesn't seem like we have used the exact machinery within this repo to publish as on npm `graphql-esm` is up to 16.5 and the tag is just 'latest' not 'latest-esm'.

So, where are we with this? Some options:

(1) we can get rid of the build qualifiers and tag suffix, and have a package name suffix of "-esm" and every time we publish to `graphql`, we should also publish to `graphql-esm`....
(2) we can get rid of the cjs version and also get rid of the build qualifiers and tag suffix.

So either way, the end goal will be to get rid of the build qualifier and tag suffix, right?

But have we decided on what will happen?
